geopandas/geodataframe.py

```diff
@@ -1691,12 +1691,12 @@ def explode(self, column=None, ignore_index=False, index_parts=None, **kwargs):
             )
             exploded_index = exploded_geom.index
 
-        df = (
-            df_copy.drop(df_copy._geometry_column_name, axis=1)
-            .join(exploded_geom)
-            .set_geometry(self._geometry_column_name)
-            .__finalize__(self)
-        )
+        df = GeoDataFrame(
+            df_copy.drop(df_copy._geometry_column_name, axis=1).loc[
+                exploded_geom.index
+            ],
```

Comment:
Hmm, next gotcha :) 
I think this won't work with if you have duplicate values in the index (which you ideally don't have, but still, it is allowed):

```python
from shapely.geometry import MultiPoint

df = GeoDataFrame(
    {"vals": [1, 2, 3]},
    geometry=[MultiPoint([(x, x), (x, 0)]) for x in range(3)],
    index=[1, 1, 2],
)

In [4]: df.explode()
<ipython-input-4-bc4232b19bee>:1: FutureWarning: Currently, index_parts defaults to True, but in the future, it will ...
ValueError: cannot reindex from a duplicate axis
```

Now, this actually also fails on the main branch (just with a different error), so it's not a regression (and thus could also be addressed later).

Reply:
I suppose the general solution to this is doing the join / take based on positional numbers (eg from a default 0, 1, 2..n range index).

Might be worth looking at how pandas' DataFrame.explode does this, because they basically do something similar as us here (exploding on one columns, and then joining the other columns with this result): https://github.com/pandas-dev/pandas/blob/4ce84de5250aafa706d4c7610fdefad0baf37acf/pandas/core/frame.py#L8334-L8351 
The strange thing is that it doesn't seem to try to restore an original index (in the beginning it does `df = self.reset_index(drop=True)` without storing the original index) ..

Reply:
Right. Third attempt.

Reply:
Just a note on the non unique index - this will now fix #2229 as well, and I'll close #2262 correspondingly.

Reply:
Whoops, sorry for forgetting about https://github.com/geopandas/geopandas/pull/2262

Reply:
That's all good, this is cleaner anyway I think. It makes sense to try and catch these cases at once if it simplifies the solution.